hepdata/modules/records/api.py

```diff
@@ -1043,3 +1043,139 @@ def get_all_ids(index=None, id_field='recid', last_updated=None, latest_first=Fa
+
+    :return: [list] A list of HEPSubmission objects
+    """
+    related_submissions = []
+    for related in submission.related_recids:
+        data_submission = (
+            HEPSubmission.query
+            .filter(HEPSubmission.publication_recid == related.related_recid)
+            .filter(HEPSubmission.overall_status == 'finished').first()
+        )
+        if data_submission:
+            related_submissions.append(data_submission)
+    return related_submissions
+
+
+def get_related_to_this_hepsubmissions(submission):
+    """
+    Queries the database for all records in the RelatedRecId table
+    that have THIS record's id as a related record.
+    Then returns the HEPSubmission object marked in the RelatedRecid table.
+    Returns only submissions marked as 'finished'
+
+    :return: [list] List containing related records.
+    """
+    related_submissions = (
+        HEPSubmission.query
+        .join(RelatedRecid, RelatedRecid.this_recid == HEPSubmission.publication_recid)
+        .filter(RelatedRecid.related_recid == submission.publication_recid)
+        .filter(HEPSubmission.overall_status == 'finished')  # Only finished submissions
+        .all()
+    )
+    return related_submissions
+
+
+def get_related_datasubmissions(data_submission):
+    """
+    Queries the database for all DataSubmission objects contained in
+    this object's related DOI list.
+    Only returns an object if associated HEPSubmission status is 'finished'
+    (All submissions this one is relating to)
+
+    :param data_submission: The datasubmission object to find related data for.
+    :return: [list] A list of DataSubmission objects
+    """
+    related_submissions = []
+    for related in data_submission.related_tables:
+        submission = (
+            DataSubmission.query.filter(DataSubmission.doi == related.related_doi)
+                      .join(HEPSubmission, HEPSubmission.publication_recid == DataSubmission.publication_recid)
+                      .filter(HEPSubmission.overall_status == 'finished')
+                      .first()
+        )
+        if submission:
+            related_submissions.append(submission)
+    return related_submissions
+
+
+def get_related_to_this_datasubmissions(data_submission):
+    """
+        Get the DataSubmission Objects with a RelatedTable entry
+        where this doi is referred to in related_doi.
+
+        :param data_submission: The datasubmission to find the related entries for.
+        :return: [List] List of DataSubmission objects.
+    """
+    related_submissions = (
+        DataSubmission.query.join(RelatedTable, RelatedTable.table_doi == DataSubmission.doi)
+        .join(HEPSubmission,
+              (HEPSubmission.publication_recid == DataSubmission.publication_recid)
+              & (HEPSubmission.version == DataSubmission.version))
+        .filter(RelatedTable.related_doi == data_submission.doi)
+        .filter(HEPSubmission.overall_status == 'finished')
+        .all()
```

Comment:
Check that the `HEPSubmission` object corresponding to the `table_doi` is the latest finished version in case there are more than one.  Links where the `table_doi` is from an earlier version of a record should not be displayed.